src/services/factories/tokenList.ts

```diff
@@ -93,27 +103,119 @@ export function getTokensFactory(factoryParams: {
+    // Fetch addresses from contract given numTokens count
+    const addressesAndIds = await retry({
+      fn: fetchAddressesAndIds,
+      fnParams: [networkId, numTokens],
+    })
+    logDebug(`[${networkId}] Token id and address mapping:`)
+    addressesAndIds.forEach((id, address) => logDebug(`[${networkId}] ${id} : ${address}`))
+
+    // Create a map of current token list addresses and tokens
+    const localAddressesMap = new Map<string, TokenDetails>(
+      tokenListApi
+        // Get a copy of the current token list
+        .getTokens(networkId)
+        // Remove tokens that are on the list but not registered on the contract
+        .filter(({ address }) => addressesAndIds.has(address))
+        // Map it with token address as key for easy access
+        .map(token => [token.address, token]),
+    )
+
+    const promises: Promise<TokenFromErc20 | string>[] = []
+
+    // Go over all value (id) and key (tokenAddress) pairs registered on the contract
+    addressesAndIds.forEach((id, tokenAddress) => {
+      if (!localAddressesMap.has(tokenAddress)) {
+        // New token not in our local list, fetch erc20 details for it
+        logDebug(`[${networkId}] Address ${tokenAddress} with id ${id} not in local list, fetching`)
+        promises.push(getErc20DetailsOrAddress(networkId, tokenAddress))
+      } else {
+        // Token already exists, update id
+        logDebug(`[${networkId}] Address ${tokenAddress} already in the list, updating id to ${id}`)
+        const token = localAddressesMap.get(tokenAddress) as TokenDetails
+        token.id = id
+      }
+    })
+
+    const partialTokens = await Promise.all(promises)
+
+    // For all newly fetched tokens
+    partialTokens.forEach(partialToken => {
+      if (typeof partialToken === 'string') {
+        // We replaced potential null responses with original tokenAddress string for logging purposes
+        logDebug(`[${networkId}] Address ${partialToken} is not a valid ERC20 token`)
+      } else if (partialToken) {
+        // If we got a valid response
+        logDebug(
+          `[${networkId}] Got details for address ${partialToken.address}: symbol '${partialToken.symbol}' name '${partialToken.name}'`,
+        )
+        // Get id from address/id mapping
+        const id = addressesAndIds.get(partialToken.address) as number
+        // build token object
+        const token: TokenDetails = { ...partialToken, id }
+        // add to local address map
+        localAddressesMap.set(partialToken.address, token)
+      }
+    })
+
+    // Convert map values to a list. Map keeps insertion order, so new tokens are added at the end
+    const tokenList = Array.from(localAddressesMap.values())
+
+    // Persist it \o/
+    tokenListApi.persistTokens({ networkId, tokenList })
+  }
+
+  async function updateTokens(networkId: number): Promise<void> {
+    updatedTokensForNetwork.add(networkId)
```

Comment:
should we update `updatedTokensForNetwork` first thing? or should we do it once they are updated?

How would it behave with a concurrent get, for example when it's awaiting in line 196

Reply:
I'm doing as a first thing to avoid having multiple update calls in parallel.

Consider the situation:
1. getTokens is called, updateTokens is triggered async and tokens are returned to the caller.
1. getTokens is called again. updateTokens haven't yet finished, flag is false, updateTokens is executed again, tokens are returned to the caller...

And so on. Makes sense?

Reply:
Make sense, although second call can return the same promise if it's updating?

```js
if (updatingPromise) {
  return updatingPromise 
} else {
  updatingPromise = update()
}
```

Reply:
So you rather have a separated state for tracking when it's being updated, besides the state where we track when it had successfully updated, or have the former instead of the later?

Both are amounting to very similar results in my mind, unless you mean something different and I'm not following?